File: mechanical/src/main/java/resonantinduction/mechanical/fluid/pipe/RenderPipe.java
package resonantinduction.mechanical.fluid.pipe;

import java.awt.Color;

import net.minecraft.item.ItemDye;
import net.minecraft.item.ItemStack;
import net.minecraft.util.ResourceLocation;
import net.minecraftforge.common.ForgeDirection;
import net.minecraftforge.fluids.FluidStack;

import org.lwjgl.opengl.GL11;

import codechicken.lib.colour.Colour;
import codechicken.lib.colour.ColourARGB;
import resonantinduction.core.Reference;
import calclavia.lib.render.FluidRenderUtility;
import calclavia.lib.render.RenderUtility;
import calclavia.lib.render.item.ISimpleItemRenderer;
import calclavia.lib.utility.WorldUtility;
import cpw.mods.fml.relauncher.Side;
import cpw.mods.fml.relauncher.SideOnly;

@SideOnly(Side.CLIENT)
public class RenderPipe implements ISimpleItemRenderer
{
	public static final RenderPipe INSTANCE = new RenderPipe();

	public static ModelPipe MODEL_PIPE = new ModelPipe();
	public static ResourceLocation TEXTURE = new ResourceLocation(Reference.DOMAIN, Reference.MODEL_PATH + "pipe.png");

	public void render(PartPipe part, double x, double y, double z, float f)
	{
		GL11.glPushMatrix();
		GL11.glTranslatef((float) x + 0.5F, (float) y + 1.5F, (float) z + 0.5F);
		GL11.glScalef(1.0F, -1F, -1F);

		if (part.getColor() > 0)
		{
			Color insulationColour = new Color(ItemDye.dyeColors[part.getColor()]);
			GL11.glColor4f(insulationColour.getRed() / 255f, insulationColour.getGreen() / 255f, insulationColour.getBlue() / 255f, 1);
		}
		else
		{
			GL11.glColor4f(1, 1, 1, 1);
		}

		render(0, part.getAllCurrentConnections());
		GL11.glPopMatrix();

		GL11.glPushMatrix();

		FluidStack fluid = part.getPressureTank().getFluid();
		int capacity = part.getPressureTank().getCapacity();
		byte renderSides = part.getAllCurrentConnections();

		if (fluid != null && fluid.amount > 0)
		{
			float percentage = (float) fluid.amount / (float) capacity;
			int[] displayList = FluidRenderUtility.getFluidDisplayLists(fluid, part.world(), false);
			RenderUtility.bind(FluidRenderUtility.getFluidSheet(fluid));
			Color color = new Color(fluid.getFluid().getColor());
			GL11.glColor4f(color.getRed() / 255f, color.getGreen() / 255f, color.getBlue() / 255f, fluid.getFluid().isGaseous() ? 0.5f : 1);

			GL11.glPushMatrix();
			GL11.glPushAttrib(GL11.GL_ENABLE_BIT);
			GL11.glEnable(GL11.GL_CULL_FACE);
			GL11.glDisable(GL11.GL_LIGHTING);
			GL11.glEnable(GL11.GL_BLEND);
			GL11.glBlendFunc(GL11.GL_SRC_ALPHA, GL11.GL_ONE_MINUS_SRC_ALPHA);

			GL11.glTranslatef((float) x + 0.35f, (float) y + 0.35f, (float) z + 0.35f);
			GL11.glScalef(0.3f, 0.3f, 0.3f);

			GL11.glCallList(displayList[(int) (percentage * (FluidRenderUtility.DISPLAY_STAGES - 1))]);

			GL11.glPopAttrib();
			GL11.glPopMatrix();

			for (ForgeDirection direction : ForgeDirection.VALID_DIRECTIONS)
			{
				if (WorldUtility.isEnabledSide(renderSides, direction))
				{
					GL11.glPushMatrix();
					GL11.glPushAttrib(GL11.GL_ENABLE_BIT);
					GL11.glEnable(GL11.GL_CULL_FACE);
					GL11.glDisable(GL11.GL_LIGHTING);
					GL11.glEnable(GL11.GL_BLEND);
					GL11.glBlendFunc(GL11.GL_SRC_ALPHA, GL11.GL_ONE_MINUS_SRC_ALPHA);

					switch (direction.ordinal())
					{
						case 0:
							GL11.glTranslatef((float) x + 0.35F, (float) y - 0f, (float) z + 0.35F);
							break;
						case 1:
							GL11.glTranslatef((float) x + 0.35F, (float) y + 0.65f, (float) z + 0.35F);
							break;
						case 2:
							GL11.glTranslatef((float) x + 0.35F, (float) y + 0.35f, (float) z + 0F);
							break;
						case 3:
							GL11.glTranslatef((float) x + 0.35F, (float) y + 0.35f, (float) z + 0.65F);
							break;
						case 4:
							GL11.glTranslatef((float) x + 0F, (float) y + 0.35f, (float) z + 0.35F);
							break;
						case 5:
							GL11.glTranslatef((float) x + 0.65F, (float) y + 0.35f, (float) z + 0.35F);
							break;
					}

					GL11.glScalef(0.3f, 0.3f, 0.3f);
					GL11.glCallList(displayList[(int) (percentage * (FluidRenderUtility.DISPLAY_STAGES - 1))]);

					GL11.glPopAttrib();
					GL11.glPopMatrix();
				}
			}
		}
		GL11.glPopMatrix();

	}

	public static void render(int meta, byte sides)
	{
		RenderUtility.enableBlending();
		RenderUtility.bind(TEXTURE);
		MODEL_PIPE.render(sides);
		RenderUtility.disableBlending();
	}

	@Override
	public void renderInventoryItem(ItemStack itemStack)
	{
		GL11.glPushMatrix();
		GL11.glTranslatef(0.5F, 1.5F, 0.5F);
		GL11.glScalef(1.0F, -1F, -1F);
		render(itemStack.getItemDamage(), Byte.parseByte("001100", 2));
		GL11.glPopMatrix();
	}
}

File: src/main/scala/edx/mechanical/fluid/pipe/RenderPipe.scala
package edx.mechanical.fluid.pipe

import java.awt.Color

import cpw.mods.fml.relauncher.{Side, SideOnly}
import edx.core.Reference
import net.minecraft.item.{ItemDye, ItemStack}
import net.minecraft.util.ResourceLocation
import net.minecraftforge.client.IItemRenderer.ItemRenderType
import net.minecraftforge.client.model.AdvancedModelLoader
import net.minecraftforge.common.util.ForgeDirection
import org.lwjgl.opengl.GL11
import resonantengine.api.item.ISimpleItemRenderer
import resonantengine.lib.render.{FluidRenderUtility, RenderUtility}
import resonantengine.lib.transform.vector.Vector3
import resonantengine.lib.utility.WorldUtility
import resonantengine.lib.wrapper.BitmaskWrapper._

@SideOnly(Side.CLIENT)
object RenderPipe extends ISimpleItemRenderer
{
  val model = AdvancedModelLoader.loadModel(new ResourceLocation(Reference.domain, Reference.modelPath + "pipe.tcn"))
  var texture = new ResourceLocation(Reference.domain, Reference.modelPath + "pipe.png")

  /**
   * Render pipe in world
   */
  def render(part: PartPipe, x: Double, y: Double, z: Double, f: Float)
  {
    GL11.glPushMatrix()
    GL11.glTranslated(x + 0.5, y + 0.5, z + 0.5)
    render(part.getMaterialID, if (part.getColor > 0) ItemDye.field_150922_c(part.getColor) else -1, part.clientRenderMask)

    val fluid = part.node.getFluid
    val pos = new Vector3(x, y, z)

    if (fluid != null && fluid.amount > 0)
    {
      ForgeDirection.VALID_DIRECTIONS.filter(d => part.clientRenderMask.mask(d.ordinal())).foreach(
        dir =>
        {
          GL11.glPushMatrix()
          GL11.glTranslated(dir.offsetX * 0.33, dir.offsetY * 0.33, dir.offsetZ * 0.33)
          GL11.glScaled(0.33, 0.33, 0.33)
          val tank = part.node
          FluidRenderUtility.renderFluidTesselation(tank, 1, 1, 1, 1)
          GL11.glPopMatrix()
        })
    }

    GL11.glPopMatrix()
  }

  /**
   * Render inventory pipe
   */
  def renderInventoryItem(renderType: ItemRenderType, itemStack: ItemStack, data: AnyRef*)
  {
    GL11.glPushMatrix()

    if (renderType == ItemRenderType.EQUIPPED_FIRST_PERSON || renderType == ItemRenderType.EQUIPPED)
      GL11.glTranslated(0.5, 0.5, 0.5)

    render(itemStack.getItemDamage, -1, 0xC)
    GL11.glPopMatrix()
  }

  /**
   * Render Pipe Model
   */
  def render(meta: Int, colorCode: Int, sides: Int)
  {
    RenderUtility.enableBlending()
    RenderUtility.bind(texture)
    val material = PipeMaterials(meta).asInstanceOf[PipeMaterials.PipeMaterial]
    val matColor = new Color(material.color)

    GL11.glColor4f(matColor.getRed / 255f, matColor.getGreen / 255f, matColor.getBlue / 255f, 1)

    model.renderOnly("Mid")

    for (dir <- ForgeDirection.VALID_DIRECTIONS)
    {
      if (WorldUtility.isEnabledSide(sides, dir))
      {
        GL11.glColor4f(matColor.getRed / 255f, matColor.getGreen / 255f, matColor.getBlue / 255f, 1)
        var prefix: String = null

        dir match
        {
          case ForgeDirection.DOWN =>
            prefix = "Bottom"
          case ForgeDirection.UP =>
            prefix = "Top"
          case ForgeDirection.NORTH =>
            prefix = "Front"
          case ForgeDirection.SOUTH =>
            prefix = "Back"
          case ForgeDirection.WEST =>
            prefix = "Right"
          case ForgeDirection.EAST =>
            prefix = "Left"
          case _ =>
        }

        model.renderOnly(prefix + "Inter", prefix + "Connect")

        if (colorCode > 0)
        {
          val color = new Color(colorCode)
          GL11.glColor4f(color.getRed / 255f, color.getGreen / 255f, color.getBlue / 255f, 1)
        }

        model.renderOnly(prefix + "Pipe")
      }
    }

    RenderUtility.disableBlending()
  }
}